File: bindings/ruby/examples/smoke.rb
# frozen_string_literal: true

# End-to-end smoke test for the Ruby bindings.
#
# Run: ruby -Ilib examples/smoke.rb

$LOAD_PATH.unshift File.expand_path("../lib", __dir__)
require "rockbox_ffi"

REPO = File.expand_path("../../..", __dir__)
FIXTURE = File.join(REPO, "crates", "rocksky", "fixtures", "08 - Internet Money - Speak(Explicit).m4a")

puts "ABI version: #{RockboxFFI.abi_version}"

# -- metadata -----------------------------------------------------------
meta = RockboxFFI::Metadata.read(FIXTURE)
puts "codec=#{meta[:codec].inspect} title=#{meta[:title].inspect} " \
     "artist=#{meta[:artist].inspect} duration_ms=#{meta[:duration_ms]} " \
     "sample_rate=#{meta[:sample_rate]}"
raise "codec label should not be empty" if meta[:codec].to_s.empty?

probe = RockboxFFI::Metadata.probe("song.flac")
puts "probe('song.flac') = #{probe.inspect}"
raise "expected FLAC, got #{probe.inspect}" unless probe == "FLAC"

# -- DSP: -6.0206 dB track gain should HALVE amplitude ------------------
rate = 44_100
RockboxFFI::Dsp.open(rate) do |dsp|
  dsp.set_replaygain(RockboxFFI::DspReplayGainMode::TRACK, false, 0.0)
  dsp.set_replaygain_gains(track_gain_db: -6.0206) # x0.5

  sine = RockboxFFI.sine_stereo(1000, 1.0, rate, 16_000)
  out = dsp.process(sine)
  peak = out.map(&:abs).max
  puts "DSP peak after -6.02 dB track gain: #{peak} (expected ~8000)"
  raise "peak #{peak} out of range" unless (7_600..8_400).cover?(peak)
end

# -- Player (construct only, no audible playback) -----------------------
RockboxFFI::Player.open(volume: 0.0) do |player|
  sr = player.sample_rate
  puts "Player sample_rate=#{sr}"
  raise "sample_rate should be > 0" unless sr.positive?

  player.set_volume(0.0)
  player.set_queue([FIXTURE])
  sleep 0.1 # the queue command is applied asynchronously by the engine thread

  st = player.status
  puts "Player status: state=#{st[:state].inspect} queue_len=#{st[:queue_len]}"
  raise "expected stopped, got #{st[:state].inspect}" unless st[:state] == "stopped"
  raise "expected queue_len 1, got #{st[:queue_len]}" unless st[:queue_len] == 1
end

puts "\nALL CHECKS PASSED ✔"

File: bindings/python/examples/smoke.py
"""End-to-end smoke test for the Python bindings.

Run: `uv run python examples/smoke.py`
"""

from __future__ import annotations

import time
from pathlib import Path

import rockbox_ffi as rb
from rockbox_ffi import Dsp, Player, metadata
from rockbox_ffi.enums import DspReplayGainMode, ReplayGainMode

REPO = Path(__file__).resolve().parents[3]
FIXTURE = REPO / "crates" / "rocksky" / "fixtures" / "08 - Internet Money - Speak(Explicit).m4a"


def main() -> None:
    print(f"ABI version: {rb.abi_version()}")

    # -- metadata ---------------------------------------------------------
    meta = metadata.read(str(FIXTURE))
    print(f"codec={meta['codec']!r} title={meta['title']!r} "
          f"artist={meta['artist']!r} duration_ms={meta['duration_ms']} "
          f"sample_rate={meta['sample_rate']}")
    assert meta["codec"], "codec label should not be empty"

    probe = metadata.probe("song.flac")
    print(f"probe('song.flac') = {probe!r}")
    assert probe, "flac probe should return a label"

    # -- DSP: -6.0206 dB track gain should HALVE amplitude ----------------
    rate = 44100
    with Dsp(rate) as dsp:
        dsp.set_replaygain(DspReplayGainMode.TRACK, noclip=False, preamp_db=0.0)
        dsp.set_replaygain_gains(track_gain_db=-6.0206)  # x0.5

        sine = rb.sine_stereo(1000.0, 1.0, rate, amplitude=16000)
        out = dsp.process(sine)
        peak = max(abs(s) for s in out)
        print(f"DSP peak after -6.02 dB track gain: {peak} (expected ~8000)")
        assert 7600 <= peak <= 8400, f"peak {peak} out of range"

    # -- Player (construct only, no audible playback) ---------------------
    with Player(volume=0.0) as player:
        rate = player.sample_rate()
        print(f"Player sample_rate={rate}")
        assert rate > 0
        player.set_volume(0.0)
        player.set_queue([str(FIXTURE)])
        time.sleep(0.1)  # the queue command is applied asynchronously by the engine thread
        st = player.status()
        print(f"Player status: state={st['state']!r} queue_len={st['queue_len']}")
        assert st["state"] == "stopped"
        assert st["queue_len"] == 1

    print("\nALL CHECKS PASSED ✔")


if __name__ == "__main__":
    main()
